Additional functionalities to further demo Audience in action (#272)
microsoft/FluidExamples: brainstorm/README.md

    - With our newly generated and updated list of new notes, we call `setNotes` to update the React state. This updated React state will propagate the changes to all remote clients, resulting in the view updating.

## Using Audience to Render User Information

> The LetsBrainstorm app make use of the `audience` property from `FrsContainerServices` to keep track of and render all user related information.

Reviewer: ```suggestion
The LetsBrainstorm app makes use of the `audience` property from `FrsContainerServices` to keep track of and render all user related information.
```

To sync the data, we created a `setMembersCallback()` function, which retrieves a list of all the active members and convert it to an array, then have a listener keep listening for the "membersChanged" event, and fire the function each time. Now React will handle updating the view each time the new `members` state is modified.

> Now, audience also has a `getMyself()` property to get the current client as a member. Passing this into the view as props will allows the user information to be displayed or processed when the user performs different note operations (creating a note, liking a note, and editing a note).

Author: I think we all agree that this README is a bit verbose, especially in the audience section. After discussing with @heliocliu, I'm going to take out the explanation for all audience features and instead just focus on the "last edited by" as an walk-through example, similar to `sharedmap` in the previous section.

```ts
const authorInfo = audience.getMyself();
```

With `members` and `authorInfo` defined, we can used these to achieve couple of tasks:

1. displaying all current active users
2. displaying author name in persona tooltip 
3. displaying like and the note's liked users
4. displaying the note's last edited user

### 1. Displaying All Current Actuve Users
To display all active members in the session, we are using the `Facepile` element from `@fluentui/react`. This element will display all the active members in their respective personas. The persona will display the image if the user has one, otherwise, it will display the first letter of their names. Now, in order to use the `Facepile` element, we first pass the `member` variable into our `Header` element, defined [here](./src/view/Header.tsx), as argument.

> ```ts

Reviewer 2: I read through the full README but it's really verbose, not sure how useful this is as a learning tool.  I also note that although this gives a blow-by-blow of the code, there are few to none code comments reflecting similar info.  Perhaps consider significantly reducing the README in favor of in-code documentation of the bits that matter most.

Author: I also feel like it's very verbose, but only in the audience section. I decided to take out explaining all audience features and instead just dive into the "last edited by" feature as a walk-through example

To sync the data, we created a `setMembersCallback()` function, which retrieves a list of all the active members and convert it to an array, then have a listener keep listening for the "membersChanged" event, and fire the function each time. Now React will handle updating the view each time the new `members` state is modified.

> The audience object also has a `getMyself()` function that returns the current client as a member. This is passed in as a view prop so that the user information to be displayed or processed when the user performs different note operations (creating a note, liking a note, and editing a note).

Reviewer 3: ```suggestion
The audience object also has a `getMyself()` function that returns the current client as a member. This is passed in as a view prop so that the user information can be displayed or processed when the user performs different note operations (creating a note, liking a note, and editing a note).
```

```ts
const authorInfo = audience.getMyself();
```

With `members` and `authorInfo` defined, we can use these to achieve several tasks:

1. displaying all current active users
2. displaying author name in persona tooltip 

> 3. displaying like and the note's liked users

Reviewer 3: ```suggestion
3. displaying likes for the note
```

4. displaying the note's last edited user

### Example Walk-through
Because the usage of the `audience` objects work in a similar fashion, let's focus on the more complex use case, editing a note and displaying the note's last edited user. When displaying the last edited user for the note, we are taking into account the current and the last edited user. If the last edited user is the same as the current user, instead of displaying the user's name, we display "Last edited by you" to be more intuitive. It is also important to define that only when the user alters the content/text inside the body of a note is it considered editing. In other words, only when `SetNoteText()` in [BrainstormModel](./src/BrainstormModel.ts) is called will we update the note's last edited user.

```ts
const setText = (text: string) => {
    model.SetNoteText(note.id, text, props.author);
};
```

> As seen in here in [NoteSpace.tsx](./src/view/NoteSpace.tsx), the `setText()` function calls the `SetNoteText()` function from [BrainstormModel](./src/BrainstormModel.ts), passing in `props.author`, which is the `authorInfo` passed to the `NoteSpace` element as argument. Similar to `onLike()` mentioned previously, the `setText()` function defined here is passed into [NoteBody.tsx](./src/view/NoteBody.tsx) as props where it will be called when the text changes as seen below.

Reviewer 3: ```suggestion
As seen here and in [NoteSpace.tsx](./src/view/NoteSpace.tsx), the `setText()` function calls the `SetNoteText()` function from [BrainstormModel](./src/BrainstormModel.ts), passing in `props.author`, which is the `authorInfo` passed to the `NoteSpace` element. The `setText()` function is passed into [NoteBody.tsx](./src/view/NoteBody.tsx) where it will be called when the text changes as seen below.
```

```ts
return (
    <div style={{ flex: 1 }}>
      <TextField
        styles={{ fieldGroup: { background: ColorOptions[color].light } }}
        borderless
        multiline
        resizable={false}
        autoAdjustHeight
        onChange={(event) => setText(event.currentTarget.value)}
        value={text}
        placeholder={"Enter Text Here"}
      />
    </div>
  );
```

Going back to `SetNoteText()` from [BrainstormModel](./src/BrainstormModel.ts), we can see in the definition below that we are not only updating the last edited member but also giving it a timestamp of when it was last edited.

```ts
const SetNoteText = (noteId: string, noteText: string, lastEditedMember: FrsMember) => {
    sharedMap.set(c_TextPrefix + noteId, noteText);
    sharedMap.set(c_LastEditedPrefix + noteId, { member: lastEditedMember, time: Date.now() });
  };
```

> The reason for adding a timestamp is because given that Fluid updates so quickly, in the event where multiple users are editing the same note, we want to wait a little bit after all changes are done before displaying the last edited user. By doing this, we can prevent the text field from flickering with user names, which could potentially be distracting.

Reviewer 3: ```suggestion
In the event where multiple users are editing the same note, we want to wait a little bit after all changes are done before displaying the last edited user. By doing this, we can prevent the text field from flickering with user names.
```
